lib/src/main/kotlin/com/swmansion/starknet/data/serializers/TransactionSerializer.kt

```diff
@@ -0,0 +1,91 @@
+package com.swmansion.starknet.data.serializers
+
+import com.swmansion.starknet.data.types.*
+import kotlinx.serialization.KSerializer
+import kotlinx.serialization.SerializationException
+import kotlinx.serialization.descriptors.PrimitiveKind
+import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
+import kotlinx.serialization.encoding.Decoder
+import kotlinx.serialization.encoding.Encoder
+import kotlinx.serialization.json.*
+
+internal object TransactionSerializer : KSerializer<Transaction> {
+    override val descriptor = PrimitiveSerialDescriptor("Transaction", PrimitiveKind.STRING)
+
+    override fun deserialize(decoder: Decoder): Transaction {
+        require(decoder is JsonDecoder)
+        val element = decoder.decodeJsonElement()
+        val jsonElement = element.jsonObject
+        val typeElement = jsonElement.getOrElse("type") { throw SerializationException("Input element does not contain mandatory field 'type'") }
+
+        val type = decoder.json.decodeFromJsonElement(TransactionType.serializer(), typeElement)
+
+        return when (type) {
+            TransactionType.INVOKE -> deserializeInvoke(decoder, element)
+            TransactionType.DECLARE -> deserializeDeclare(decoder, element)
+            TransactionType.DEPLOY_ACCOUNT -> deserializeDeployAccount(decoder, element)
+            TransactionType.DEPLOY -> Json.decodeFromJsonElement(DeployTransaction.serializer(), element)
+            TransactionType.L1_HANDLER -> Json.decodeFromJsonElement(L1HandlerTransaction.serializer(), element)
+        }
+    }
+
+    override fun serialize(encoder: Encoder, value: Transaction) {
+        require(encoder is JsonEncoder)
+
+        val jsonObject = when (value) {
+            is InvokeTransactionV3 -> Json.encodeToJsonElement(InvokeTransactionV3.serializer(), value).jsonObject
```

Comment:
I think we should be using `encoder.json.encodeToJsonElement` instead of `Json.encodeToJsonElement`, to use instance provided by the calling `encoder`, so if there are any extra settings defined at the higher level, we don't override those by using `Json` instance with default ones.